### PR TITLE
campaigns: Introduce heuristic syncing of changesets

### PR DESCRIPTION
**Pull request by ryanslade**
This changes our syncer from syncing all changesets every 2 minutes to one that syncs each changeset based on a combination of when it was last synced and when it last changed. We are also able to inject "high priority" items for immediate syncing.

For now, the heuristics used are very simply and more of a linear backoff than anything else. We can extend this later if needed.

The syncer operates in the following way:

A schedule is generated on a set interval, currently every two minutes. The schedule is computed by fetching sync data from the database:

* Last sync with codehost
* Last remote change. ie, what was the codehost "updated_at" value when we last synced
* Latest event (this could be newer than either of the above due to webhooks)

From the above, we work out when next to sync by looking at the difference between when we last synced and max(remote change, latest event) and performing linear backoff.

For example:

We cap syncs with a max 8 hours, min of 2 minutes.

Last sync 10am, latest change 9:55. Next sync 10:05.
Last sync 10am, latest change 9:50. Next sync 10:10.
Last sync 10am. latest change 10:01. Next sync 10:02 (Sync happened before latest event but we'll never sync more frequently than every 2 minutes)

Once we have the schedule we use it to trigger our syncs in chronological order.

Every two minutes the schedule is replaced. This is less efficient but we need to fetch new data from the db anyway and it simplifies the code being able to simply throw away the old schedule and use a new one.

Still TODO in followup PRs:
* Instrumentation / dashboards
* Shared, configurable rate limiter
* Sync now UI 
* Concurrency. We only process once sync at a time for now
* We could potentially calculate the next sync time in the db and return a limited subset of items that will sync soon but I've left the work in code for simplicity at this point

Closes: https://github.com/sourcegraph/sourcegraph/issues/6388

### CODE REVIEW

File: enterprise/internal/campaigns/syncer_test.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+			lastChange: clock(),
+			want:       clock().Add(minSyncDelay),
+		},
+		{
+			name:       "Linear backoff",
+			clock:      clock,
+			lastSync:   clock(),
+			lastChange: clock().Add(-1 * time.Hour),
+			want:       clock().Add(1 * time.Hour),
+		},
```
Comment (unknwon) [inline enterprise/internal/campaigns/syncer_test.go]: Maybe another test case for `lastChange: clock().Add(time.Hour),` for the sake of completeness?

**Comment by unknwon** [comment]
👍 Overall LGTM, just having a concern about capacity of the priority channel.

File: enterprise/internal/campaigns/syncer.go [hunk 2]
```diff
@@ -17,39 +19,134 @@ type ChangesetSyncer struct {
...
+
+	queue := newChangesetQueue(100)
+	// Get initial schedule
+	for {
+		sched, err := s.computeSchedule(ctx)
+		if err != nil {
+			log15.Error("Computing queue", "err", err)
+			time.Sleep(5 * time.Second)
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (ryanslade) [reply enterprise/internal/campaigns/syncer.go]: Yes, we could cancel after a few attempts as it'll continue to try in the main loop lower down anyway

File: enterprise/internal/campaigns/syncer.go [hunk 3]
```diff
@@ -17,39 +19,134 @@ type ChangesetSyncer struct {
...
+	ctx := context.Background()
+	scheduleInterval := s.ScheduleInterval
+	if scheduleInterval == 0 {
+		scheduleInterval = 2 * time.Minute
+	}
+	if s.clock == nil {
+		s.clock = time.Now
+	}
+
+	queue := newChangesetQueue(100)
```
[the remark this answers is not in the document]
Comment (ryanslade) [reply enterprise/internal/campaigns/syncer.go]: The priority queue will only be triggered by a user manually clicking "sync now" on a changeset so I don't think it needs to be too large.

File: enterprise/internal/campaigns/syncer.go [hunk 4]
```diff
@@ -215,6 +312,75 @@ func (s *ChangesetSyncer) listAllNonDeletedChangesets(ctx context.Context) (all
...
+	ctx := context.Background()
+	ctx, q.cancel = context.WithCancel(ctx)
+	var timer *time.Timer
+	go func() {
+		for i := range schedule {
+			// Get most urgent changeset and sleep until it should be synced
+			now := time.Now()
+			nextSync := schedule[i].nextSync
+			d := nextSync.Sub(now)
+			timer = time.NewTimer(d)
```
[the remark this answers is not in the document]
Comment (ryanslade) [reply enterprise/internal/campaigns/syncer.go]: I'll change that, the code originally reused a timer so that got missed in the refactor

File: enterprise/internal/campaigns/syncer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/internal/campaigns/syncer.go]: Gotcha, make sense.

File: enterprise/internal/campaigns/syncer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/internal/campaigns/syncer.go]: Oh, right... Then I think we don't even need to retry, it is most likely a database problem and we're "retrying" every X minutes.

**Comment by ryanslade** [comment]
@unknwon PTAL

File: enterprise/internal/campaigns/syncer.go [hunk 5]
```diff
@@ -215,6 +312,75 @@ func (s *ChangesetSyncer) listAllNonDeletedChangesets(ctx context.Context) (all
...
 
+type changesetQueue struct {
+	scheduled chan int64
+	priority  chan int64
+
+	// ordered slice of scheduled syncs
+	cancel context.CancelFunc
+}
+
+func newChangesetQueue(priorityCapacity int) *changesetQueue {
```
[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/internal/campaigns/syncer.go]: Thanks for the docstring, very helpful!

Another thing popped up to my head, please think about the following scenario:

A user or multiple users want to sync a changeset C, but because we're busying syncing earlier "priority" changesets, they lost patience and triggered N sync requests for changeset C. 

Now it's the term to process the very first sync request of changeset C, then what about rest (N-1) sync requests for this exact same changeset? Doing another (N-1) times sync seems pretty wasted.

In short, do we need deduplication in `scheduled` and `priority` "queues"?

PS: I kinda lost track of details if this is already addressed :)

**Comment by unknwon** [comment]
@ryanslade Thanks! 

In case you didn't notice, there are few comments "hidden" in the middle.

File: enterprise/internal/campaigns/syncer.go [hunk 6]
```diff
@@ -17,39 +19,137 @@ type ChangesetSyncer struct {
...
+		s.clock = time.Now
+	}
+
+	s.queue = newChangesetQueue(100)
+	// Get initial schedule
+	attempts := 0
+	maxAttempts := 5
+	for attempts < maxAttempts {
+		attempts++
+		sched, err := s.computeSchedule(ctx)
```
Comment (mrnugget) [inline enterprise/internal/campaigns/syncer.go]: Why this seemingly excessive retry-logic? If I understand the `s.computeSchedule` call correctly, the only error that might be returned here would come from the database (connection). Since we never retry database queries in any other place, why here? And might crashing & restarting `repo-updater` be a better option?

And if you keep this retry-behavior: how about moving it in `s.computeSchedule` and wrapping only the call to `ListChangesetSyncHeuristics`?

File: enterprise/internal/campaigns/syncer.go [hunk 7]
```diff
@@ -17,39 +19,137 @@ type ChangesetSyncer struct {
...
-		return err
+		return nil, errors.Wrap(err, "listing changeset heuristics")
 	}
 
-	if err := s.SyncChangesets(ctx, cs...); err != nil {
-		log15.Error("ChangesetSyncer", "error", err)
-		return err
+	ss := make([]syncSchedule, len(hs))
+	for i := range hs {
+		nextSync := nextSync(s.clock, hs[i].UpdatedAt, maxTime(hs[i].ExternalUpdatedAt, hs[i].LatestEvent))
```
Comment (mrnugget) [inline enterprise/internal/campaigns/syncer.go]: The `maxTime` here is neat!

Idea: I would change the signature of `nextSync` so that only you only call `nextSync(s.clock, hs[i])`. That way the complete "algorithm" is in `nextSync`. Having the `maxTime` here is a bit confusing.

File: enterprise/internal/campaigns/syncer.go [hunk 8]
```diff
@@ -215,6 +315,80 @@ func (s *ChangesetSyncer) listAllNonDeletedChangesets(ctx context.Context) (all
...
+				timer.Stop()
+				return
+			case <-timer.C:
+				// Timer ready, try and send sync instruction
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case q.scheduled <- schedule[i].changesetID:
```
Comment (mrnugget) [inline enterprise/internal/campaigns/syncer.go]: This is blocked for as long as it takes to `SyncChangesetByID` the scheduled ID, right? That means backpressure might build up here and in the next iteration of the loop the `d` might be negative, if we took longer than `d` to sync the previous changeset. Does the code handle a negative `d`?

File: enterprise/internal/campaigns/syncer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mrnugget) [reply enterprise/internal/campaigns/syncer.go]: Agree with the thoughts here. I wonder whether there isn't a possible approach in which `enqueuePriority` results in the given changeset landing at the top of the `[]syncSchedule` instead of a separate queue. That would solve the deduplication issue and it would also remove the need for dropping priority-requests.

File: cmd/repo-updater/repos/bitbucketserver.go [hunk 9]
```diff
@@ -202,6 +203,8 @@ func (s BitbucketServerSource) LoadChangesets(ctx context.Context, cs ...*Change
 		}
 
 		cs[i].Changeset.ExternalBranch = git.AbbreviateRef(pr.FromRef.ID)
+		// Bitbucket Server timestamps are in milliseconds
+		cs[i].Changeset.ExternalUpdatedAt = time.Unix(0, int64(pr.UpdatedDate*1000000))
```
[the remark this answers is not in the document]
Comment (ryanslade) [reply cmd/repo-updater/repos/bitbucketserver.go]: Copied as I'm not sure where to put it so that it is accessible from both packages

File: enterprise/internal/campaigns/store.go [hunk 10]
```diff
@@ -540,6 +547,56 @@ func getChangesetQuery(opts *GetChangesetOpts) *sqlf.Query {
 	return sqlf.Sprintf(getChangesetsQueryFmtstr, sqlf.Join(preds, "\n AND "))
 }
 
+// ChangesetSyncHeuristics represents data about the sync status of a changeset
+type ChangesetSyncHeuristics struct {
```
[the remark this answers is not in the document]
Comment (ryanslade) [reply enterprise/internal/campaigns/store.go]: Nope, will do

File: enterprise/internal/campaigns/syncer.go [hunk 11]
```diff
@@ -17,39 +19,137 @@ type ChangesetSyncer struct {
...
+// and is expected to be launched once at startup.
+func (s *ChangesetSyncer) StartSyncing() {
+	// TODO: Setup instrumentation here
+	ctx := context.Background()
+	scheduleInterval := s.ScheduleInterval
+	if scheduleInterval == 0 {
+		scheduleInterval = 2 * time.Minute
+	}
+	if s.clock == nil {
+		s.clock = time.Now
```
[the remark this answers is not in the document]
Comment (ryanslade) [reply enterprise/internal/campaigns/syncer.go]: I'm not sure I understand. I'm setting default values if none were provided. I do set a default for scheduleInterval above

File: enterprise/internal/campaigns/syncer.go (hunk 6 above)
Comment (ryanslade) [reply enterprise/internal/campaigns/syncer.go]: I removed the retry logic. We currently rechedule every two minutes anyway which is probably frequent enough to act as a retry anyway

File: cmd/repo-updater/repos/github.go [hunk 12]
```diff
@@ -224,6 +224,7 @@ func (s GithubSource) LoadChangesets(ctx context.Context, cs ...*Changeset) erro
 	for i := range cs {
 		cs[i].Changeset.ExternalBranch = prs[i].HeadRefName
 		cs[i].Changeset.Metadata = prs[i]
+		cs[i].Changeset.ExternalUpdatedAt = prs[i].UpdatedAt
```
[the remark this answers is not in the document]
Comment (ryanslade) [reply cmd/repo-updater/repos/github.go]: I'd rather not modify the meaning of the field. I think it should always be whatever the code host thinks is the latest update time. 

The scheduler uses the difference between their value and our sync time to decide when next to sync. If we set ExternalUpdatedAt to be the same as UpdatedAt then the scheduler will think the item just changed and will do another sync shortly.

File: enterprise/internal/campaigns/syncer.go (hunk 8 above)
Comment (ryanslade) [reply enterprise/internal/campaigns/syncer.go]: Negative d is fine, it just means that the sync loop won't sleep and will try to sync immediately as it's behind schedule.

File: enterprise/internal/campaigns/store.go [hunk 13]
```diff
@@ -540,6 +547,56 @@ func getChangesetQuery(opts *GetChangesetOpts) *sqlf.Query {
...
+
+func listChangesetSyncHeuristicsQuery() *sqlf.Query {
+	return sqlf.Sprintf(`
+SELECT changesets.id,
+       changesets.updated_at,
+       max(ce.updated_at) as latest_event,
+       changesets.external_updated_at
+FROM changesets
+JOIN changeset_events ce ON changesets.id = ce.changeset_id
+GROUP by changesets.id
```
[the remark this answers is not in the document]
Comment (ryanslade) [reply enterprise/internal/campaigns/store.go]: hah, I've actually edited my editor config now so that it enforces uppercase

File: enterprise/internal/campaigns/store.go [hunk 14]
```diff
@@ -540,6 +547,56 @@ func getChangesetQuery(opts *GetChangesetOpts) *sqlf.Query {
...
+}
+
+func listChangesetSyncHeuristicsQuery() *sqlf.Query {
+	return sqlf.Sprintf(`
+SELECT changesets.id,
+       changesets.updated_at,
+       max(ce.updated_at) as latest_event,
+       changesets.external_updated_at
+FROM changesets
+JOIN changeset_events ce ON changesets.id = ce.changeset_id
```
[the remark this answers is not in the document]
Comment (ryanslade) [reply enterprise/internal/campaigns/store.go]: Ah, good point

**Comment by ryanslade** [comment]
@mrnugget @unknwon I've dealt with all changes apart from the priority scheduling concerns as I'd like to do that in another PR to make things clearer

File: enterprise/internal/campaigns/store.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/internal/campaigns/store.go]: > edited my editor config now so that it enforces uppercase

Mind sharing it? I WANT IT!

**Comment by unknwon** [comment]
This file seems having some comments/suggestions that are not addressed? https://github.com/sourcegraph/sourcegraph/pull/8653/files#diff-02ffa3baa5a42c3c6898b51c72d79514

**Comment by ryanslade** [comment]
> This file seems having some comments/suggestions that are not addressed? https://github.com/sourcegraph/sourcegraph/pull/8653/files#diff-02ffa3baa5a42c3c6898b51c72d79514

Oops, not sure how I missed those. I'll take a look now

File: cmd/repo-updater/repos/github.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/repo-updater/repos/github.go]: > If we set ExternalUpdatedAt to be the same as UpdatedAt then the scheduler will think the item just changed and will do another sync shortly.

No, that's not what I meant. I solely meant to ask: do we handle "external clock skew"? But after looking at the rest of the code I think we're fine here.

File: enterprise/internal/campaigns/syncer.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (mrnugget) [reply enterprise/internal/campaigns/syncer.go]: Yeah, but for one you set the default value _on_ `s` (`s.clock`) and for the other one you only set the local variable `scheduleInterval` (as opposed to `s.ScheduleInterval`). I think my confusion comes from the fact that this method acts kinda like a constructor but also kinda not :)

File: enterprise/internal/campaigns/syncer.go (hunk 8 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/syncer.go]: Okay, thanks!

File: enterprise/internal/campaigns/syncer.go [hunk 15]
```diff
@@ -17,39 +19,129 @@ type ChangesetSyncer struct {
...
-// database
-func (s *ChangesetSyncer) Sync(ctx context.Context) error {
-	cs, err := s.listAllNonDeletedChangesets(ctx)
+var (
+	minSyncDelay = 2 * time.Minute
+	maxSyncDelay = 8 * time.Hour
+)
+
+// nextSync computes the time we want the next sync to happen.
+func nextSync(h campaigns.ChangesetSyncHeuristics) time.Time {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/syncer.go]: If the name `ChangesetSyncHeuristics` sticks (see my other comment), maybe this should be a method on it?

File: enterprise/internal/campaigns/syncer.go [hunk 16]
```diff
@@ -215,6 +307,80 @@ func (s *ChangesetSyncer) listAllNonDeletedChangesets(ctx context.Context) (all
...
+			}
+		}
+	}()
+}
+
+func (q *changesetQueue) enqueuePriority(ids []int64) error {
+	// q.priority will be buffered so that we allow
+	// a fixed depth in the high priority queue and
+	// can perform a non blocking add here
+	for _, id := range ids {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/syncer.go]: How about validating that `len(ids) < priorityCapacity`? But then again, from a caller's perspective it's always possible, when `len(ids) > 1`, that _some_ of the IDs made it and others don't and you can't tell.

**Comment by ryanslade** [comment]
@mrnugget @unknwon PT(Y)AL

**Comment by ryanslade** [comment]
I'm going to merge this and will send a follow up PR to rework the scheduler to better handle high priority items as mentioned

File: enterprise/internal/campaigns/syncer.go [hunk 17]
```diff
@@ -17,39 +19,129 @@ type ChangesetSyncer struct {
...
+		ss[i] = syncSchedule{
+			changesetID: hs[i].ChangesetID,
+			nextSync:    nextSync,
+		}
 	}
-	return nil
+
+	// This will happen in the db later, for now we'll grab everything and order in code
+	sort.Slice(ss, func(i, j int) bool {
+		return ss[i].nextSync.Before(ss[j].nextSync)
```
Comment (keegancsmith) [inline enterprise/internal/campaigns/syncer.go]: you want to tie break here, so we get some sort of determinism. The natural thing to use is the changesetID

File: enterprise/internal/campaigns/syncer.go [hunk 18]
```diff
@@ -215,6 +307,80 @@ func (s *ChangesetSyncer) listAllNonDeletedChangesets(ctx context.Context) (all
...
+	if q.cancel != nil {
+		// cancel closes the done chan on the existing context
+		// which will cause the existing worker routine to exit
+		q.cancel()
+	}
+
+	ctx := context.Background()
+	ctx, q.cancel = context.WithCancel(ctx)
+	go func() {
+		for i := range schedule {
```
Comment (keegancsmith) [inline enterprise/internal/campaigns/syncer.go]: nit: why not loop over the values?

File: enterprise/internal/campaigns/syncer.go [hunk 19]
```diff
@@ -215,6 +307,80 @@ func (s *ChangesetSyncer) listAllNonDeletedChangesets(ctx context.Context) (all
...
+			nextSync := schedule[i].nextSync
+			d := nextSync.Sub(now)
+			timer := time.NewTimer(d)
+			select {
+			case <-ctx.Done():
+				timer.Stop()
+				return
+			case <-timer.C:
+				// Timer ready, try and send sync instruction
+			}
```
Comment (keegancsmith) [inline enterprise/internal/campaigns/syncer.go]: nit: we use a mini helper in the repos scheduler to make this easier to read `func sleep(ctx context.Context, d time.Duration)`. You can even make the helper immediately return if `d <= 0`.
